File: datafn/python/datafn/logger.py
from typing import Any, Dict, Optional, Protocol, runtime_checkable


@runtime_checkable
class DatafnLogger(Protocol):
    def info(self, message: str, context: Optional[Dict[str, Any]] = None) -> None: ...
    def warn(self, message: str, context: Optional[Dict[str, Any]] = None) -> None: ...
    def error(self, message: str, context: Optional[Dict[str, Any]] = None) -> None: ...
    def debug(self, message: str, context: Optional[Dict[str, Any]] = None) -> None: ...


class NoopLogger:
    def info(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        pass

    def warn(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        pass

    def error(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        pass

    def debug(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        pass


class ConsoleLogger:
    def info(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        print(f"[INFO] {message}", end="")
        if context:
            print(f" {context}", end="")
        print()

    def warn(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        print(f"[WARN] {message}", end="")
        if context:
            print(f" {context}", end="")
        print()

    def error(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        print(f"[ERROR] {message}", end="")
        if context:
            print(f" {context}", end="")
        print()

    def debug(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        print(f"[DEBUG] {message}", end="")
        if context:
            print(f" {context}", end="")
        print()


def create_logger(logger: Any = None) -> DatafnLogger:
    if logger is not None:
        return logger
    return NoopLogger()

File: plugfn/core/src/utils/logger.ts
import type { Logger } from '../types/action.js';

/**
 * Console logger implementation
 */
export class ConsoleLogger implements Logger {
  constructor(private prefix = '[PlugFn]') {}

  debug(message: string, meta?: any): void {
    console.debug(`${this.prefix} DEBUG:`, message, meta || '');
  }

  info(message: string, meta?: any): void {
    console.info(`${this.prefix} INFO:`, message, meta || '');
  }

  warn(message: string, meta?: any): void {
    console.warn(`${this.prefix} WARN:`, message, meta || '');
  }

  error(message: string, meta?: any): void {
    console.error(`${this.prefix} ERROR:`, message, meta || '');
  }
}

/**
 * No-op logger that doesn't output anything
 */
export class NoopLogger implements Logger {
  debug(): void {}
  info(): void {}
  warn(): void {}
  error(): void {}
}

/**
 * Create a logger with a specific prefix
 */
export function createLogger(prefix: string): Logger {
  return new ConsoleLogger(prefix);
}

